sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -113,7 +113,14 @@ abstract class Expression extends TreeNode[Expression] {
 
   def nullable: Boolean
 
-  def references: AttributeSet = AttributeSet.fromAttributeSets(children.map(_.references))
+  /**
+   * Workaround scala compiler so that we can call super on lazy vals
+   */
+  @transient
+  private lazy val _references: AttributeSet =
```

Comment:
Sorry to be late to the party, but can we please change to lazy references + a compute method? It seems a bit dirty that we are overriding references everywhere and the we are adding lazy vals all over the place.

Reply:
Because somehow the scala compiler won't allow us to call super on lazy vals. This workaround makes subclasses' `reference` able to call `super.references`.

Reply:
Where do you want to call super?

Reply:
Git grep gives me `sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:    if (plan.resolved) super.references -- plan.outputSet else super.references`. This is the case requiring this workaround.

Reply:
A related Scala forum discussion: https://www.scala-lang.org/old/node/11315